File: scripts/create-user.js
#!/usr/bin/env node
/**
 * Prints the SQL that creates (or updates) a dashboard account.
 *
 *   node scripts/create-user.js <username> <password> ["Full Name"] [role]
 *
 * There is deliberately no sign-up endpoint — accounts are made by hand
 * and pasted into the Supabase SQL Editor. That way nothing on the public
 * internet can create a login for the dashboard.
 *
 * The password is hashed here, on your machine; only the digest is
 * printed, so the plaintext never reaches Supabase or the SQL logs.
 */
const { hashPassword } = require('../lib/auth');

const [, , username, password, fullName = '', role = 'admin'] = process.argv;

if (!username || !password) {
  console.error('Usage: node scripts/create-user.js <username> <password> ["Full Name"] [admin|staff]');
  process.exit(1);
}
if (password.length < 8) {
  console.error('Password must be at least 8 characters.');
  process.exit(1);
}
if (role !== 'admin' && role !== 'staff') {
  console.error("Role must be 'admin' or 'staff'.");
  process.exit(1);
}

const q = (s) => "'" + String(s).replace(/'/g, "''") + "'";

console.log(`
-- Run this in the Supabase SQL Editor.
-- Re-running with a new password resets that account's password.
insert into dashboard_users (username, password_hash, full_name, role)
values (${q(username.trim().toLowerCase())}, ${q(hashPassword(password))}, ${q(fullName)}, ${q(role)})
on conflict (lower(username)) do update
  set password_hash   = excluded.password_hash,
      full_name       = excluded.full_name,
      role            = excluded.role,
      is_active       = true,
      failed_attempts = 0,
      locked_until    = null;
`.trim());

File: scripts/create_user.py
#!/usr/bin/env python3
"""Prints the SQL that creates (or updates) a dashboard account.

    python scripts/create_user.py <username> <password> ["Full Name"] [admin|staff]

Same output as scripts/create-user.js — this exists because the machine
this was built on has Python but no Node. Python's hashlib.scrypt with
n=16384, r=8, p=1 is byte-identical to Node's crypto.scryptSync defaults,
which is what lib/auth.js verifies against.

There is deliberately no sign-up endpoint — accounts are made by hand and
pasted into the Supabase SQL Editor, so nothing on the public internet can
mint a login for the dashboard. The password is hashed here, on your
machine; only the digest is printed, so the plaintext never reaches
Supabase or its query logs.
"""
import base64
import hashlib
import os
import sys


def hash_password(password: str) -> str:
    """Returns "scrypt$<salt-b64>$<hash-b64>" — the shape stored in the DB."""
    salt = os.urandom(16)
    digest = hashlib.scrypt(
        password.encode("utf-8"), salt=salt, n=16384, r=8, p=1, dklen=64)
    return "scrypt${}${}".format(
        base64.b64encode(salt).decode(), base64.b64encode(digest).decode())


def q(value: str) -> str:
    return "'" + str(value).replace("'", "''") + "'"


def main(argv):
    if len(argv) < 3:
        sys.exit('Usage: python scripts/create_user.py <username> <password> '
                 '["Full Name"] [admin|staff]')

    username, password = argv[1], argv[2]
    full_name = argv[3] if len(argv) > 3 else ''
    role      = argv[4] if len(argv) > 4 else 'admin'

    if len(password) < 8:
        sys.exit('Password must be at least 8 characters.')
    if role not in ('admin', 'staff'):
        sys.exit("Role must be 'admin' or 'staff'.")

    print("""
-- Run this in the Supabase SQL Editor.
-- Re-running with a new password resets that account's password.
insert into dashboard_users (username, password_hash, full_name, role)
values ({}, {}, {}, {})
on conflict (lower(username)) do update
  set password_hash   = excluded.password_hash,
      full_name       = excluded.full_name,
      role            = excluded.role,
      is_active       = true,
      failed_attempts = 0,
      locked_until    = null;
""".format(q(username.strip().lower()), q(hash_password(password)),
           q(full_name), q(role)).strip())


if __name__ == '__main__':
    main(sys.argv)
